### PR TITLE
Integrate eslint with eslint-config-standard

### LINKED ISSUE #366: Add eslint configuration
This is an idea from https://github.com/TypeFox/monaco-languageclient/issues/315

There currently is no eslint configuration. We should introduce one.

### PR DESCRIPTION
**Pull request by kaisalmen**
I have based this branch on your branch, because otherwise the changes are obsolete again soon and would need to be redone. Therefore the target branch is currently also `use-monaco-vscode-api`

@CGNonofr I used `eslint-config-standard`, but I made some adjustments:
- Indent is 4 instead of 2
- I favor semi-colons
- `no-dupe-class-member` is off
- `no-redeclare`, `no-useless-constructor` and `no-void` have been reduced to **warn** (you should check if the code should or could be changed or if `eslint-disable` should be set instead and we can then remove the rule reduction)

This closes #366

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        }],
+        'space-before-function-paren': ['error', {
+            anonymous: 'always',
+            named: 'never',
+            asyncArrow: 'always'
+        }],
+        semi: ['error', 'always'],
+        'no-extra-semi': 'off',
+        '@typescript-eslint/no-extra-semi': ['error'],
+        'no-dupe-class-members': 'off',
```
Comment (CGNonofr) [inline .eslintrc.js:27]: That's the base javascript rule that doesn't work well with typescript, what about enabling `@typescript-eslint/no-dupe-class-members`?

File: .eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        'no-redeclare': 'warn',
```
Comment (CGNonofr) [inline .eslintrc.js:28]: Same, what about enabling `@typescript-eslint/no-redeclare`?

File: .github/workflows/actions.yml [hunk 3]
```diff
@@ -17,4 +17,6 @@ jobs:
     - name: Build
       shell: bash
       run: |
-        npm i && npm run build
+        npm i
```
Comment (CGNonofr) [inline .github/workflows/actions.yml:20]: Why is it `npm i` instead of `npm ci`?

File: packages/client/src/monaco-language-client.ts [hunk 4]
```diff
@@ -52,28 +54,28 @@ export class MonacoLanguageClient extends BaseLanguageClient {
...
                 ...params,
                 registrations: params.registrations.filter(registration => this.getFeature(<any>registration.method) != null)
-            })
-        }
-        const originalHandleUnregistrationRequest: (params: UnregistrationParams) => Promise<void> = this['handleUnregistrationRequest'].bind(this)
-        this['handleUnregistrationRequest'] = (params: UnregistrationParams) => {
+            });
+        };
+        const originalHandleUnregistrationRequest: (params: UnregistrationParams) => Promise<void> = this['handleRegistrationRequest'].bind(this);
+        this['handleRegistrationRequest'] = (params: UnregistrationParams) => {
```
Comment (CGNonofr) [inline packages/client/src/monaco-language-client.ts:65]: I don't think it should be changed

**Comment by CGNonofr** [comment]
Btw, you can use the `header/header` rule to ensure all files have the proper header

File: .eslintrc.js (hunk 2 above)
Comment (kaisalmen) [reply .eslintrc.js:28]: Same as above. Thanks 🙂

File: .github/workflows/actions.yml (hunk 3 above)
Comment (kaisalmen) [reply .github/workflows/actions.yml:20]: Good point

File: packages/client/src/monaco-language-client.ts (hunk 4 above)
Comment (kaisalmen) [reply packages/client/src/monaco-language-client.ts:65]: Sorry, stupid copy paste issue, when I changed the code back (manually)

**Comment by kaisalmen** [comment]
I implemented the review comments and rebased the branch. Now, the build will fail because `vscode-ws-jsonrpc` will have lint related errors. I will fix this afternoon, I think.

**Comment by kaisalmen** [comment]
I have now enabled eslint globally for all packages and thanks for mentioning `header/header`. I have now integrated it.

**Comment by kaisalmen** [comment]
Build is green. Everything is in. If you agree with the latest changes this is good to go, I think.

File: .github/workflows/actions.yml [hunk 5]
```diff
@@ -17,4 +17,6 @@ jobs:
     - name: Build
       shell: bash
       run: |
-        npm i && npm run build
+        npm ci
+        npm run build
+        npm run lint
```
Comment (CGNonofr) [inline .github/workflows/actions.yml:22]: Isn't it different? what happens if the npm ci fails?

File: packages/client/src/disposable.ts [hunk 6]
```diff
@@ -3,31 +3,29 @@
...
+import { Disposable } from 'vscode-jsonrpc';
 
 export class DisposableCollection implements Disposable {
     protected readonly disposables: Disposable[] = [];
 
-    dispose(): void {
+    dispose (): void {
         while (this.disposables.length !== 0) {
-            this.disposables.pop()!.dispose();
+            this.disposables.pop()?.dispose();
```
Comment (CGNonofr) [inline packages/client/src/disposable.ts:13]: What was the issue with the `!`?

File: packages/client/src/monaco-language-client.ts [hunk 7]
```diff
@@ -2,81 +2,83 @@
  * Copyright (c) 2018-2022 TypeFox GmbH (http://www.typefox.io). All rights reserved.
  * Licensed under the MIT License. See License.txt in the project root for license information.
  * ------------------------------------------------------------------------------------------ */
+
+/* eslint-disable dot-notation */
```
Comment (CGNonofr) [inline packages/client/src/monaco-language-client.ts:6]: Are you sure it's a good idea to disable a rule for a whole file?

File: packages/examples/browser-lsp/src/serverWorker.ts [hunk 8]
```diff
@@ -67,22 +66,21 @@ function getColorInformation(textDocument: TextDocumentIdentifier) {
     return colorInfos;
 }
 
-function getColorPresentation(color: Color, range: Range) {
+function getColorPresentation (color: Color, range: Range) {
     const result: ColorPresentation[] = [];
-    const red256 = Math.round(color.red * 255), green256 = Math.round(color.green * 255), blue256 = Math.round(color.blue * 255);
+    const red256 = Math.round(color.red * 255); const green256 = Math.round(color.green * 255); const blue256 = Math.round(color.blue * 255);
```
Comment (CGNonofr) [inline packages/examples/browser-lsp/src/serverWorker.ts:71]: I'm surprised there is no eslint rule preventing putting multiple variables on the same line

File: packages/vscode-ws-jsonrpc/src/disposable.ts [hunk 9]
```diff
@@ -1,33 +1,33 @@
...
                     disposables.splice(index, 1);
                 }
             }
-        }
+        };
     }
-
-}
\ No newline at end of file
+}
```
Comment (CGNonofr) [inline packages/vscode-ws-jsonrpc/src/disposable.ts:33]: (not related to this PR) Are you sure it's relevant to have the same code duplicated in the client and vscode-ws-jsonrpc?

File: packages/client/src/disposable.ts (hunk 6 above)
Comment (kaisalmen) [reply packages/client/src/disposable.ts:13]: None, my error, happend after first round of lint issue resolution. I will change back.

File: .github/workflows/actions.yml (hunk 5 above)
Comment (CGNonofr) [reply .github/workflows/actions.yml:22]: Nice to know :+1:

File: packages/examples/browser-lsp/src/serverWorker.ts (hunk 8 above)
Comment (kaisalmen) [reply packages/examples/browser-lsp/src/serverWorker.ts:71]: No, it should not be there. I changed the config to this now (upcoming commit)
```javascript
        'dot-notation': 'off',
        '@typescript-eslint/dot-notation': ['error']
```
but now VSCode is auto-formatting the code to dotted notation all the time and this leads to TypeScript errors, for example:

```shell
src/monaco-language-client.ts:55:103 - error TS2341: Property 'handleRegistrationRequest' is private and only accessible within class 'BaseLanguageClient'.

55         const originalHandleRegistrationRequest: (params: RegistrationParams) => Promise<void> = this.handleRegistrationRequest.bind(this);
```

File: packages/client/src/monaco-language-client.ts (hunk 7 above)
Comment (kaisalmen) [reply packages/client/src/monaco-language-client.ts:6]: No, it should not be there. I changed the config to this now (upcoming commit)
```javascript
        'dot-notation': 'off',
        '@typescript-eslint/dot-notation': ['error']
```
but now VSCode is auto-formatting the code to dotted notation all the time and this leads to TypeScript errors, for example:

```shell
src/monaco-language-client.ts:55:103 - error TS2341: Property 'handleRegistrationRequest' is private and only accessible within class 'BaseLanguageClient'.

55         const originalHandleRegistrationRequest: (params: RegistrationParams) => Promise<void> = this.handleRegistrationRequest.bind(this);
```

Comment (CGNonofr) [reply packages/client/src/monaco-language-client.ts:6]: Yes, this is a known hack to access a private field in typescript!

The best way is to disable the dot-notation error on each line I think

File: packages/examples/browser-lsp/src/serverWorker.ts (hunk 8 above)
Comment (kaisalmen) [reply packages/examples/browser-lsp/src/serverWorker.ts:71]: At least it is not configured in `standard`. 🙂

File: packages/client/src/monaco-language-client.ts (hunk 7 above)
Comment (kaisalmen) [reply packages/client/src/monaco-language-client.ts]: Did this now. See new commit

File: packages/vscode-ws-jsonrpc/src/disposable.ts (hunk 9 above)
Comment (kaisalmen) [reply packages/vscode-ws-jsonrpc/src/disposable.ts:33]: Not nice, but both packages are independent. Only the examples use both.

Comment (kaisalmen) [reply packages/vscode-ws-jsonrpc/src/disposable.ts:33]: I actually have a good idea for that.

Comment (kaisalmen) [reply packages/vscode-ws-jsonrpc/src/disposable.ts:29]: I forced another update. At least both `disposable.ts` are identical now.

**Comment by CGNonofr** [comment]
> I will produce new `dev` releases once this is merged. Sounds good?

Sounds good! What will we wait before releasing a final version?

**Review by CGNonofr — APPROVED** [review]
LGTM

**Comment by kaisalmen** [comment]
> Sounds good! What will we wait for before releasing a final version?

Actually a good question. We don't really have official beta tester. And a major version change should not immediately cripple correctly configured projects depending on here.

**Comment by CGNonofr** [comment]
It seems pretty safe to release a major version to me, and release fixes later, or we won't get any feedback

I'm already testing the new vscode-api approch for more than a week, and everything else is just refactors, it can't break things, can it?

**Comment by kaisalmen** [comment]
> it can't break things, can it?

Let's go

**Comment by kaisalmen** [comment]
Done 🎉 

https://www.npmjs.com/package/monaco-languageclient/v/2.0.0

https://www.npmjs.com/package/vscode-ws-jsonrpc/v/1.0.0

**Comment by kaisalmen** [comment]
We should archive the repos of `@codingame/monaco-jsonrpc` and `vscode-ws-jsonrpc`  and add links in the README that link to here.

**Comment by CGNonofr** [comment]
done!

**Comment by kaisalmen** [comment]
✅

**Pull request merged**